### PR TITLE
[SYCL] Allow kernel names in anonymous namespace

### PR DESCRIPTION
**Pull request by srividya-sundaram**
Originally, the Clang FE code rejected kernel name types declared within an anonymous namespace because those types would not be forward declarable at the global namespace scope.

However, based on a discussion between the SYCL spec authors and the Khronos group about what is meant by 'at namespace scope', it was decided that phrase means "at the same namespace as the original declaration", and so anonymous namespaces are acceptable because we can forward declare into an anonymous namespace.

Because of this relaxation in forward declaration of non-inline kernel names, it is now legal to declare a kernel name in the anonymous namespace.

### CODE REVIEW

**Comment by AaronBallman** [comment]
Can you add a description to the PR that explains why this change is happening? (This helps when we need to do a `git blame` for some reason.)

**Comment by erichkeane** [comment]
So the allowing of anonymous namespaces is NOT because of the int-footer solution, it is that we CAN forward declare anonymous namespaces (and probably should be!).  There are some solutions we can do now that that will allow this, such as making sure we preserve anonymous namespaces when forward declaring types.

**Review by erichkeane — COMMENTED** [review]
Quick look seems OK, but needs some integration-header type tests to make sure we're forward declaring correctly.

**Comment by srividya-sundaram** [comment]
> Quick look seems OK, but needs some integration-header type tests to make sure we're forward declaring correctly.

Included a test case for kernel name in anonymous namespace in clang/test/CodeGenSYCL/int_header1.cpp
Is that sufficient? 
If not, can you elaborate what needs to be specifically checked for anonymous case?

**Comment by erichkeane** [comment]
> 
> 
> > Quick look seems OK, but needs some integration-header type tests to make sure we're forward declaring correctly.
> 
> Included a test case for kernel name in anonymous namespace in
> [clang/test/CodeGenSYCL/int_header1.cpp](https://github.com/intel/llvm/pull/3649/files#diff-916ffd5ae51393639f1951d6ef2dd81e7ec94c676386b7ab5d962c7a2e5b182eR119)
> Is that sufficient?
> If not, can you elaborate what needs to be specifically checked for anonymous case?

Thats most of the way (and the part I was more worried about), but we also need to check to make sure that the forward declaration in the int-header is also in an anonymous namespace.  I don't see any such tests in that test-file however, which is also concerning...

File: clang/test/CodeGenSYCL/int_header1.cpp [hunk 1]
```diff
@@ -1,6 +1,12 @@
 // RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -sycl-std=2020 -fsycl-int-header=%t.h %s -o %t.out
 // RUN: FileCheck -input-file=%t.h %s
 
+// Check if forward declarations of kernel names in anonymous namespace are in
```
Comment (erichkeane) [inline clang/test/CodeGenSYCL/int_header1.cpp:4]: Can you change the 2nd and 3rd to 'CHECK-NEXT' to make sure here?   Also, are we putting comments on our close braces?  if so, I'd like those added to the check line as well.

Comment (srividya-sundaram) [reply clang/test/CodeGenSYCL/int_header1.cpp:4]: > Also, are we putting comments on our close braces? if so, I'd like those added to the check line as well.
There aren't any comments on the close braces in the integration header generated.

**Comment by srividya-sundaram** [comment]
> So the allowing of anonymous namespaces is NOT because of the int-footer solution, it is that we CAN forward declare anonymous namespaces (and probably should be!). There are some solutions we can do now that that will allow this, such as making sure we preserve anonymous namespaces when forward declaring types.

OK. My understanding was different from the email thread.
Updated description with your comment.

**Comment by AaronBallman** [comment]
> > So the allowing of anonymous namespaces is NOT because of the int-footer solution, it is that we CAN forward declare anonymous namespaces (and probably should be!). There are some solutions we can do now that that will allow this, such as making sure we preserve anonymous namespaces when forward declaring types.
> 
> OK. My understanding was different from the email thread.
> Updated description with your comment.

I think the description should be self-contained rather than having people's comments replicated in it (that gets harder to understand and makes for a pretty long commit message). How about:
```
Originally, the Clang FE code rejected kernel name types declared within an anonymous namespace because those types would not be forward declarable at the global namespace scope.

However, based on a discussion between the SYCL spec authors and the Khronos group about what is meant by 'at namespace scope', it was decided that phrase means "at the same namespace as the original declaration", and so anonymous namespaces are acceptable because we can forward declare into an anonymous namespace.

Because of this relaxation in forward declaration of non-inline kernel names, it is now legal to declare a kernel name in the anonymous namespace.
```
WDYT? (@erichkeane -- please make sure I'm not misrepresenting the decision from the email thread)

**Comment by erichkeane** [comment]
> 
> 
> > > So the allowing of anonymous namespaces is NOT because of the int-footer solution, it is that we CAN forward declare anonymous namespaces (and probably should be!). There are some solutions we can do now that that will allow this, such as making sure we preserve anonymous namespaces when forward declaring types.
> > 
> > 
> > OK. My understanding was different from the email thread.
> > Updated description with your comment.
> 
> I think the description should be self-contained rather than having people's comments replicated in it (that gets harder to understand and makes for a pretty long commit message). How about:
> 
> ```
> Originally, the Clang FE code rejected kernel name types declared within an anonymous namespace because those types would not be forward declarable at the global namespace scope.
> 
> However, based on a discussion between the SYCL spec authors and the Khronos group about what is meant by 'at namespace scope', it was decided that phrase means "at the same namespace as the original declaration", and so anonymous namespaces are acceptable because we can forward declare into an anonymous namespace.
> 
> Because of this relaxation in forward declaration of non-inline kernel names, it is now legal to declare a kernel name in the anonymous namespace.
> ```
> 
> WDYT? (@erichkeane -- please make sure I'm not misrepresenting the decision from the email thread)

I like that wording, and it properly captures the discussion.

**Review by elizabethandrews — COMMENTED** [review]
Changes LGTM other than test nit mentioned by other reviewers. Please add the tests requested for namespace hierarchy. Thanks!

**Comment by srividya-sundaram** [comment]
Updated test with anonymous test case at intel/llvm-test-suite: https://github.com/intel/llvm-test-suite/pull/262

**Review by AaronBallman — APPROVED** [review]
LGTM

**Review by erichkeane — APPROVED** [review]
@vladimirlaz : I think by now @AaronBallman  and @srividya-sundaram  are the experts here :)  That said, i looked over this quickly and think it seems ok.

**Comment by AaronBallman** [comment]
This is awaiting review from @intel/llvm-reviewers-runtime, I believe.

**Review by s-kanaev — APPROVED** [review]
SYCL RT test looks good.

**Pull request merged**